### ISSUE TITLE
Cannot unselect option element when preserveSelected is true

### ISSUE THREAD

**Issue by debueb** [truckingsim/Ajax-Bootstrap-Select#85]
Hi @truckingsim,

I am unable to unselect an option element in a multiple select when the `preserveSelected` option is true. It only works after I do an ajax search.

The issue is reproducible: https://jsfiddle.net/a3ybhne4/2/

I have found the problem to be [line 630](https://github.com/truckingsim/Ajax-Bootstrap-Select/blob/master/dist/js/ajax-bootstrap-select.js#L630)

It works when I replace line 630
`that.replaceOptions(that.cacheGet(that.plugin.query));`
by 
`that.replaceOptions(that.selected);`

I am not sure why `that.plugin.query` is there in the first place and I am wondering if this change has any side effects...

Thanks for taking a look.

**Comment by truckingsim** [comment]
The `that.cacheGet(that.plugin.query)` is our caching mechanism. It's always used even if you have caching off. What cache off does is it loads the cached data and does an ajax request anyways.  With caching on, if you've already done a search for that query (whatever the query is), it will grab the cached data this time **won't** do an ajax request.

What's happening is we aren't updating the cache on change.  We set the cache value for no query on [line 605](https://github.com/truckingsim/Ajax-Bootstrap-Select/blob/master/dist/js/ajax-bootstrap-select.js#L605) but after updating [that.selected](https://github.com/truckingsim/Ajax-Bootstrap-Select/blob/master/dist/js/ajax-bootstrap-select.js#L621-L628) we never update the cache.  So then it just keeps using the cached value.

So the fix is to update the cache.  I think if we just add a `that.setCache` it will break preserveSelected so it will take some testing to make sure we don't break that fixing this.

If you feel you can make a PR with a fix for the cache I can test it to make sure it doesn't break anything else.  Otherwise it will probably be this weekend before I can get to this.

**Comment by truckingsim** [comment]
Also @markcarver I believe you wrote this caching and preserveSelected code as well if I got any of that information wrong let me know.  I'm going off of memory with it.

**Comment by markcarver** [comment]
Yes, but it's been ages since I've actually looked at any of this code TBH. I really cannot say one way or the other, however it sounds right.

**Comment by truckingsim** [comment]
@markcarver yeah I haven't looked at it much either in a long time. Thanks for confirming though!

**Comment by jengjeng** [comment]
I have fixed this issue by clean the cache object with code
`
$('.select-picker').selectpicker().ajaxSelectPicker()
// Clear the cache object
$('.select-picker').data('AjaxBootstrapSelect').list.cache = {}
`
